src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -160,6 +163,51 @@
     return aclSqlCols;
   }
 
+  /**
+   * Attempt to extract ColumnTypes from the supplied Map
+   * of column names to java.sql.Type
+   *
+   * @param typeMap Map of column names to Integer java.sql.Types
+   */
+  void addColumnTypes(Map<String, Integer> typeMap) {
+    for (Map.Entry<String, Integer> entry : typeMap.entrySet()) {
+      if (types.get(entry.getKey()) != ColumnType.UNKNOWN) {
+        continue;
+      }
+      ColumnType type;
+      switch (entry.getValue()) {
+        case Types.INTEGER:
+        case Types.SMALLINT:
+        case Types.TINYINT:
+          type = ColumnType.INT;
+          break;
+        case Types.BIGINT:
+          type = ColumnType.LONG;
+          break;
+        case Types.CHAR:
+        case Types.VARCHAR:
+        case Types.LONGVARCHAR:
```

Comment:
I think we should include the NCHAR variations and DATALINK (an URL, and the spec says getString/setString are supported).